idb/postgres/postgres_migrations.go

```diff
@@ -342,70 +382,296 @@ func m5accountCumulativeRewardsUpdate(db *IndexerDb, state *MigrationState) erro
+		}
+
+		if accounting.AppOptOutTxn(stxn) {
+			result.appLocal[txn.AssetID] = updateClose(result.appLocal[txn.AssetID], txn.Round)
+		}
+	}
+
+	// Genesis accounts could have this property
+	if numTxn == 0 {
+		result.account.created.Valid = true
+		result.account.created.Int64 = 0
+	}
+
+	return nil, result
+}
+
+// m5RewardsAndDatesPart2UpdateAccounts loops through the provided accounts and generates a bunch of updates in a
+// single transactional commit. These queries are written so that they can run in the background.
+//
+// For each account we run several queries:
+// 1. updateTotalRewards            - conditionally update the total rewards if the account wasn't closed during iteration.
+// 2. setCreateCloseAccount      - set the accounts create/close rounds.
+// 3. setCreateCloseAsset        - set the accounts created assets create/close rounds.
+// 4. setCreateCloseAssetHolding - (upsert) set the accounts asset holding create/close rounds.
+// 5. setCreateCloseApp          - set the accounts created apps create/close rounds.
+// 6. setCreateCloseAppLocal     - (upsert) set the accounts local apps create/close rounds.
+//
+// Note: These queries only work if closed_at was reset before the migration is started. That is true
+//       for the initial migration, but if we need to reuse it in the future we'll need to fix the queries
+//       or redo the query.
+func m5RewardsAndDatesPart2UpdateAccounts(db *IndexerDb, state *MigrationState, accounts []string, finalBatch bool) error {
+	var err error
 	tx, err := db.db.Begin()
 	if err != nil {
-		return fmt.Errorf("%s: tx begin: %v", cumulativeRewardsUpdateErr, err)
+		return fmt.Errorf("%s: tx begin: %v", rewardsCreateCloseUpdateErr, err)
 	}
 	defer tx.Rollback() // ignored if .Commit() first
 
-	setCumulativeReward, err := tx.Prepare(`UPDATE account SET rewardstotal = $1 WHERE addr = $2`)
+	// 1. updateTotalRewards            - conditionally update the total rewards if the account wasn't closed during iteration.
+	// $3 is the round after which new blocks will have the closed_at field set.
+	// We only set rewards_total when closed_at was set before that round.
+	updateTotalRewards, err := tx.Prepare(`UPDATE account SET rewards_total = coalesce(rewards_total, 0) + $2 WHERE addr = $1 AND coalesce(closed_at, 0) < $3`)
 	if err != nil {
-		return fmt.Errorf("%s: set rewards prepare: %v", cumulativeRewardsUpdateErr, err)
+		return fmt.Errorf("%s: set rewards prepare: %v", rewardsCreateCloseUpdateErr, err)
 	}
-	defer setCumulativeReward.Close()
+	defer updateTotalRewards.Close()
+
+	// 2. setCreateCloseAccount      - set the accounts create/close rounds.
+	// We always set the created_at field because it will never change.
+	// closed_at may already be set by the time the migration runs, or it might need to be cleared out.
+	setCreateCloseAccount, err := tx.Prepare(`UPDATE account SET created_at = $2, closed_at = coalesce(closed_at, $3) WHERE addr = $1`)
```

Comment:
I think there's a hole here where the account is closed at 999, migrated at 1000, and reopened at 1001 (while the migration is still running). This could wind up with an open account reporting closed at 999.
I think we can always set created_at to the earliest seen txn; but setting closed_at should be conditional on `WHERE addr = $1 AND microalgos = 0`. Probably means this has to be two statements.

Reply:
You are correct, and this was actually my initial implementation. I'm now leaving the closed_at field in the event where it was ever set. Nice catch, this is one of the subtle implementation details that took me a while to get right.